lib/Myriad/Bootstrap.pm

```diff
@@ -78,6 +80,55 @@ sub allow_modules {
     @ALLOWED_MODULES{@_} = (1) x @_;
 }
 
+sub open_pipe {
+    # Establish comms channel for child process
+    socketpair my $child_pipe, my $parent_pipe, $constant{AF_UNIX}, $constant{SOCK_STREAM}, $constant{PF_UNSPEC}
+        or die $!;
+
+    { # Unbuffered writes
+        my $old = select($child_pipe);
+        $| = 1; select($parent_pipe);
+        $| = 1; select($old);
+    }
+
+    make_pipe_nonblocking($parent_pipe);
+    make_pipe_nonblocking($child_pipe);
+
+    return ($parent_pipe, $child_pipe);
+}
+
+sub make_pipe_nonblocking {
+    my $pipe = shift;
+    my $flags = fcntl($pipe, $constant{F_GETFL}, 0)
+        or die "Can't get flags for the socket: $!\n";
+
+    $flags = fcntl($pipe, $constant{F_SETFL}, $flags | $constant{O_NONBLOCK})
+        or die "Can't set flags for the socket: $!\n";
+}
+
+sub check_messages_in_pipe {
+    my ($pipe, $on_read) = @_;
+    # See https://perldoc.perl.org/functions/select#select-RBITS,WBITS,EBITS,TIMEOUT
+    # for a better understanding
+    my $input = '';
+    my $rin = my $win = '';
+    vec($rin, fileno($pipe), 1) = 1;
+    my $ein = $rin | $win;
+
+    die $! unless defined(my $nfound = select my $rout = $rin, my $wout = $win, my $eout = $ein, 0.5);
+
+    if($nfound) {
+        my $rslt = sysread $pipe, my $buf, 4096;
+        return 0 unless $rslt;
+        $input .= $buf;
+        while($input =~ s/^[\r\n]*(.*)(?:\r\n)+//) {
```

Comment:
Is there a specific reason for this sort of protocol? Or did you choose records delimited by `\r\n` just because? That's fine. I am just asking if you have considered alternatives like a simple `\n` as delimiter, or `\0` or maybe something with a length prefix. `pack` offers a nice way to write such prefixed records:

```
> perl -e 'print pack "N/a", $ARGV[0]' fadsfga | od -tx1z
0000000 00 00 00 07 66 61 64 73 66 67 61                 >....fadsfga<
0000013
```

At the reading end you read 4 bytes interpret them with `unpack "N"` and then read that number of bytes.

Reply:
The communication between the child and the parent is limited in nature, this is only for development use, so a text based protocol is easy to debug and the `\r\n` is the defacto terminator :sweat_smile:

Reply:
It'd be this:

```suggestion
        while($input =~ s/^[\x0D\x0A]*(.*)(?:\x0D\x0A)+//) {
```

although I think a better option is to insist on the full terminator before extracting anything - we should be able to guarantee that the only time we ever see the `\x0D\x0A` sequence is at the end of the record, and that every record contains that terminator:

```
        while($input =~ s/^([^\x0D]*)\x0D\x0A//) {
```